File: frontend/src/types/character.ts
import type { FactionName } from './faction';
import type { GearSetBonus, GearType } from './gear';
import type { Quality } from './quality';

export type CharacterClass =
  | 'Guardian'
  | 'Priest'
  | 'Assassin'
  | 'Warrior'
  | 'Archer'
  | 'Mage';

export type SkillType =
  | 'Overdrive'
  | 'Ultimate Skill'
  | 'Secret Skill'
  | 'Special Skill'
  | 'Divine Skill';

export interface Skill {
  name: string;
  type?: SkillType;
  description: string;
  cooldown: number;
}

export interface TalentLevel {
  level: number;
  effect: string;
}

export interface Talent {
  name: string;
  talent_levels: TalentLevel[];
}

export type RecommendedGearSlot =
  | 'headgear'
  | 'chestplate'
  | 'bracers'
  | 'boots'
  | 'weapon'
  | 'accessory';

export interface RecommendedGear {
  headgear?: string;
  chestplate?: string;
  bracers?: string;
  boots?: string;
  weapon?: string;
  accessory?: string;
}

export interface RecommendedGearEntry {
  slot: RecommendedGearSlot;
  type: GearType;
  name: string;
}

export interface RecommendedSubclassEntry {
  name: string;
  icon: string | undefined;
  tier: number | undefined;
  className: string | undefined;
  bonuses: string[];
  effect: string | undefined;
}

export type RecommendedGearDetail = RecommendedGearEntry & {
  label: string;
  icon: string;
  slotIcon: string;
  setName: string | null;
  setBonus: GearSetBonus | null;
  quality: string | undefined;
  lore: string | undefined;
  stats: Record<string, string | number> | undefined;
};

export interface ActivatedSetBonus {
  setName: string;
  pieces: number;
  requiredPieces: number;
  description: string;
  activations: number;
}

export interface Character {
  name: string;
  title: string;
  quality: Quality;
  character_class: CharacterClass;
  factions: FactionName[];
  is_global: boolean;
  subclasses: string[];
  height: string;
  weight: string;
  lore: string;
  quote: string;
  origin: string;
  talent?: Talent | null;
  skills: Skill[];
  noble_phantasm: string;
  recommended_gear?: RecommendedGear | null;
  recommended_subclasses?: string[];
  last_updated: number;
}

File: backend/models/character.py
from enum import Enum

from pydantic import BaseModel


class Quality(str, Enum):
    SSR_EX = "SSR EX"
    SSR_PLUS = "SSR+"
    SSR = "SSR"
    SR_PLUS = "SR+"
    R = "R"
    N = "N"


class CharacterClass(str, Enum):
    GUARDIAN = "Guardian"
    PRIEST = "Priest"
    ASSASSIN = "Assassin"
    WARRIOR = "Warrior"
    ARCHER = "Archer"
    MAGE = "Mage"


class TalentLevel(BaseModel):
    level: int
    effect: str


class Talent(BaseModel):
    name: str
    talent_levels: list[TalentLevel]


class SkillType(str, Enum):
    OVERDRIVE = "Overdrive"
    ULTIMATE_SKILL = "Ultimate Skill"
    SECRET_SKILL = "Secret Skill"
    SPECIAL_SKILL = "Special Skill"
    DIVINE_SKILL = "Divine Skill"


class Skill(BaseModel):
    name: str
    type: SkillType | str | None = None
    description: str
    cooldown: int = 0


class Character(BaseModel):
    name: str
    title: str
    quality: Quality
    character_class: CharacterClass
    factions: list[str]
    is_global: bool
    subclasses: list[str]
    height: str
    weight: str
    lore: str
    quote: str
    origin: str
    talent: Talent | None = None
    skills: list[Skill]
    noble_phantasm: str
